ts/util/legend.ts

```diff
@@ -0,0 +1,159 @@
+import { qcStatuses } from "../data/qc-status";
+import { makeIcon } from "./html-utils";
+
+export class Legend {
+  private container: HTMLElement;
+  constructor() {
+    // outer container
+    this.container = document.createElement("div");
+    this.container.className =
+      "text-black block flex flex-col bg-white rounded-xl fixed border-2 border-grey-200 z-40 drop-shadow-lg";
+    this.container.id = "legend-container";
+    // header bar w/title and close window icon
+    const header = document.createElement("div");
+    header.className =
+      "flex flex-row items-center border-b-1 border-grey-200 cursor-move";
+    const title = document.createElement("span");
+    title.className = "font-sarabun font-thin text-green-200 text-24 my-2 mx-4";
+    title.innerHTML = "Legend";
+    const spacer = document.createElement("div");
+    spacer.className = "flex-auto";
+    const icon = makeIcon("xmark");
+    icon.classList.add(
+      "fa-solid",
+      "fa-xmark",
+      "text-grey-200",
+      "hover:text-green-200",
+      "text-24",
+      "m-2",
+      "mx-4",
+      "z-50",
+      "relative",
+      "cursor-pointer"
+    );
+    icon.addEventListener("click", () => this.container.remove());
+    header.appendChild(title);
+    header.appendChild(spacer);
+    header.appendChild(icon);
+    // grid of legend labels
+    const body = document.createElement("div");
+    body.className = "m-4 grid grid-rows-4 grid-flow-col gap-2";
+    for (const qcStatus of Object.values(qcStatuses)) {
+      const labelContainer = document.createElement("div");
+      labelContainer.className =
+        "flex items-center space-x-2 bg-grey-100 rounded-md font-inter font-medium p-2 text-12";
+      const icon = makeIcon(qcStatus.icon);
+      icon.classList.add("text-black");
+      const label = document.createElement("span");
+      label.innerHTML = qcStatus.label;
+      labelContainer.appendChild(icon);
+      labelContainer.appendChild(label);
+      body.appendChild(labelContainer);
+    }
+
+    // centering window, setTimeout is required for offsetHeight and offsetWidth to load
+    // TODO: fix flash introduced by timeout
+    window.setTimeout(() => {
+      this.container.style.top =
+        (window.innerHeight - this.container.offsetHeight) / 2 + "px";
+      this.container.style.left =
+        (window.innerWidth - this.container.offsetWidth) / 2 + "px";
+    }, 0);
```

Comment:
Can this be done by making it hidden when adding to the DOM, then make it visible here after positioning it?

Reply:
I can't tell, but it seems to be less frequent. Issue still persists though

Reply:
I just tried and it does solve the problem. Using "hidden" creates a different problem - the container has no height or width if it's hidden, so this doesn't calculate the position correctly. You can use "invisible" instead and both problems are solved - add "invisible" to the container when first created, and remove it after setting the position here.